lib/Dialect/TTNN/Transforms/Passes.cpp

```diff
@@ -3,20 +3,105 @@
-#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIR.h"
-#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
-#include "ttmlir/Dialect/TTIR/Transforms/Passes.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOpsTypes.h"
 
-namespace mlir::tt::ttnn {} // namespace mlir::tt::ttnn
+namespace mlir::tt::ttnn {
+#define GEN_PASS_DEF_TTNNDEALLOCATE
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h.inc"
+
+class TTNNDeallocate : public impl::TTNNDeallocateBase<TTNNDeallocate> {
+
+public:
+  using impl::TTNNDeallocateBase<TTNNDeallocate>::TTNNDeallocateBase;
+
+  Operation *getLastValueUsageOp(const LivenessBlockInfo *livenessInfo,
+                                 Value value) {
+    Operation *startOp = livenessInfo->getStartOperation(value);
+    Operation *endOp = livenessInfo->getEndOperation(value, startOp);
+    auto *opOperandIter =
+        llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
+          return opOperand.is(value);
+        });
+
+    // Keep going until we find the last usage of the tensor.
+    // For DPS stop if operand is not dps init one.
+    // For non DPS skip through as it might modify the tensor in place.
+    //
+    while (opOperandIter != endOp->getOpOperands().end() &&
+           ((isa<DestinationStyleOpInterface>(endOp) &&
+             cast<DestinationStyleOpInterface>(endOp).isDpsInit(
+                 &(*opOperandIter))) ||
+            (!isa<DestinationStyleOpInterface>(endOp) &&
+             endOp->getNumResults() > 0))) {
+      assert(endOp->getResults().size() == 1);
+      OpResult result = endOp->getResult(0);
+      endOp = livenessInfo->getEndOperation(result, endOp);
+      opOperandIter =
+          llvm::find_if(endOp->getOpOperands(), [&](OpOperand &opOperand) {
+            return opOperand.is(result);
+          });
+    }
+
+    return endOp;
+  }
+
+  void runOnOperation() final {
+    ModuleOp module = getOperation();
+    IRRewriter rewriter(&getContext());
+
+    module->walk([&](func::FuncOp func) {
+      assert(func.getBody().hasOneBlock());
+      Liveness liveness(func.getOperation());
+      const LivenessBlockInfo *livenessInfo =
+          liveness.getLiveness(&func.getBody().front());
+
+      // Handle non DPS ops which do not store function result and are used to
+      // allocate tensors. DPS ops are handled via ttnn::EmptyOp.
+      //
+      func->walk([&](Operation *op) {
+        if (isa<DestinationStyleOpInterface>(op)) {
+          return;
+        }
+
+        // Skip ops which do not have results.
+        //
+        if (op->getNumResults() == 0) {
+          return;
+        }
+
+        // Check if result is ranked tensor type.
+        //
+        if (!isa<RankedTensorType>(op->getResult(0).getType())) {
+          return;
+        }
+
+        // Handle ttnn::EmptyOp, ttnn::FullOp and ttnn::ToDeviceOp for now.
+        //
+        if (!isa<ttnn::EmptyOp>(op) && !isa<ttnn::FullOp>(op) &&
+            !isa<ttnn::ToDeviceOp>(op)) {
```

Comment:
Why is this specialized to only these ops?  Shouldn't this happen for all ops?

Reply:
ttnn::EmptyOp will cover all DPS ops. For non-DPS ops problem is we do not know which ops are allocating new tensors and which are making changes inline. For example I hit a problem with ttnn::reshape where dealloc of input tensor deallocates output as well(I need to double click on this, looks like some kind of TensorView ie multiple tensor objects sharing same storage?). We need some kind of tag on ops for memory effects but unfortunately for ttnn:EmptyOp I see we already started tagging them the wrong way. Therefore instead of fixing all that now and breaking others in parallel by allowing all ops to deallocate and then tightening it down op by op I decided to start safe but still endup deallocating majority of used memory.

Reply:
That being said, with implementation in latest iteration which goes through nonDPS ops I could remove this OP specialization and potentially leak memory only on nonDPS ops chain leading into function return, but I need to do more checks. For start add asserts everywhere in runtime for every op that we are not working with deallocated tensor.

Reply:
Aren't we doing a refactor that will make a lot of other TTNN ops non-dps though?

It's temporarily reverted:
- https://github.com/tenstorrent/tt-mlir/pull/857

It seems like we'd want to do this for all nonDps ops.

Reply:
@nsmithtt No-one said we don't want to do it for nonDPS ops. :) Its just that I needed to harden runtime with bunch of asserts, debug TTNN deallocator, and finally find and fix a bug in our runtime which made dealloc double free memory for nonDPS inline ops, and I wasn't looking to do it in one change. But here it is now. Good thing is it seems we do not need to tag ops manually for now, due to TTNN handling refcount properly for these Tensor views(tensors objects reusing same buffer). Might need to tweak my test a bit to cover new deallocs.

Reply:
I think we should implement it the proper way, i.e. generic for all nonDps ops and then use workarounds with issues tracking them and TTNN bugs filed.  I don't want to have special cases without issues tracking to fix in the future.

We should implement a similar struct as runtime https://github.com/tenstorrent/tt-mlir/blob/main/runtime/include/tt/runtime/detail/workarounds.h